### PR TITLE
[REF] web: move ControlPanel inside controllers/actions

### PR DESCRIPTION
**Pull request by ged-odoo**
This branch introduces a large-scale reorganization of the
component tree generated by the web client.  The short version is
that now, the control panel is a child of the view controller and
no longer a sibling. Graphically (and simplified), we go from
this:

                  webClient
                 /    |    \
               ...   ...  actionManager
                           /         \
                   controlPanel   viewController
                                    /       \

to this:

           	  webClient
                 /    |    \
               ...   ...  actionManager
                               |
                          viewController
                            /   |   \
          ControlPanelController
              /           \
         CPRenderer     CPModel

The motivation is that this work moves the code where it should be.
Before this commit, it was kind of weird to have code in the
controllers to render buttons outside of their root node (in
renderButtons).  Also, the action manager had to take care of
coordinating search view states between view transitions.

So, this work simplifies the code.  It also makes it easier to
extend. We see day after day that Odoo needs to take care of more
complex UI needs, and in many cases, these needs were quite
difficult to implement (we prefer spaghettis in our plates, not in
our code).  The changes in this branch should open the way to
implement these features.

For example,
- it will now be easy to add the possibility of views (for example,
  the search view or a new ControlPanel view) to add custom buttons
  (of type action or object) in the control panel.
- Another need will be to serialize/ restore the state of the
  search view across action boundaries (needed by the dashboard).
- Another example is the possibility for views to customize easily
  the presence/absence of sub menus (filters/groupbys/favorites/
  time range/...)
- Another need is an easier way for views/client action to customize
  their control panel.

This commit also contains a large rewrite of the search view (so it
is more inline with our architecture and easier to maintain).

Part of task 1893568

Co-authored-by: Aaron Bohy <aab@odoo.com>
Co-authored-by: Mathieu Duckerts-Antoine <dam@odoo.com>
Co-authored-by: Martin Geubelle <mge@odoo.com>

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by Elkasitu** [comment]
robodoo merge

**Comment by fmdl** [comment]
Hi @ged-odoo , I don't know what you work excatly, I long time ago I have give an idea you show a count of inactive field at the bottom of list (like gmail) during the search, please see my idea : https://github.com/odoo/odoo/issues/19505

File: addons/web/static/src/js/core/sorted_registry.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+     * @constructor
+     * @param {Object} [mapping] the initial data in the registry
+     */
+    init: function () {
+        var self = this;
+        this._super.apply(this, arguments);
+        this._scores = [];
+        this._initialKeys = Object.keys(this.map);
+        this._sortedKeys = this._initialKeys;
+        this.listeners.push(function () {
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/sorted_registry.js]: this is complicated. why not override the keys method?  Also, why do you keep the initial keys unsorted? and why do you use the sortedKeys variable here when you define the _sortedKeys above?

File: addons/web/static/src/js/core/sorted_registry.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            );
+        });
+    },
+
+    //--------------------------------------------------------------------------
+    // Public
+    //--------------------------------------------------------------------------
+
+    /**
+     * overide
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/sorted_registry.js]: not valid docstring. you should use `@override`. Also notice that it is not highlighted. Maybe your text editor is not properly configured

File: addons/web/static/src/js/core/sorted_registry.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+     *
+     *  Allows to attribute a score to the added key. This can be used to sort keys
+     *  using the scores.
+     *
+     * @param {string} key
+     * @param {any} value
+     * @param {number} score
+     * @returns {Registry} can be used to chain add calls.
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/sorted_registry.js]: you actually forgot to return the registry in the implementation

File: addons/web/static/src/js/core/sorted_registry.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * overide
+     *
+     *  Allows to attribute a score to the added key. This can be used to sort keys
+     *  using the scores.
+     *
+     * @param {string} key
+     * @param {any} value
+     * @param {number} score
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/sorted_registry.js]: if it is optional, it should be marked as optional

File: addons/web/static/src/js/core/sorted_registry.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+     * @returns {Registry} can be used to chain add calls.
+     */
+    add: function (key, value, score) {
+        this._scores.push({
+            key: key,
+            score: score || key,
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/sorted_registry.js]: this fallback  is weird. are you sure you want to compare strings and number?

File: addons/web/static/src/js/core/sorted_registry.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+        });
+        this._super.apply(this, arguments);
+    },
+    /**
+     * @overide
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/sorted_registry.js]: bad docstring

File: addons/web/static/src/js/core/sorted_registry.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+     * returns the object map keys sorted according to this.sortFunction
+     * returns {string[]}
```
Comment (ged-odoo) [inline addons/web/static/src/js/core/sorted_registry.js]: bad docstring (`@returns`)

**Comment by ged-odoo** [comment]
robodoo r+

**Comment by ged-odoo** [comment]
robodoo rebase-merge r+

**Comment by ged-odoo** [comment]
robodoo rebase-merge

**Comment by ged-odoo** [comment]
@d-fence not sure that robodoo is going to accept this... Should I do something?

**Pull request closed without merging**